src/backend/distributed/operations/shard_split.c

```diff
@@ -1336,20 +1345,37 @@ DropShardList(List *shardIntervalList)
 			/* get shard name */
 			char *qualifiedShardName = ConstructQualifiedShardName(shardInterval);
 
-			char storageType = shardInterval->storageType;
-			if (storageType == SHARD_STORAGE_TABLE)
+			if (DeferShardDeleteOnSplit)
 			{
-				appendStringInfo(dropQuery, DROP_REGULAR_TABLE_COMMAND,
-								 qualifiedShardName);
+				/* Log shard in pg_dist_cleanup.
+				 * Parent shards are to be dropped only on sucess after split workflow is complete,
+				 * so mark the policy as 'CLEANUP_DEFERRED_ON_SUCCESS'.
+				 * We also log cleanup record in the current transaction. If the current transaction rolls back,
+				 * we do not generate a record at all.
+				 */
+				CleanupPolicy policy = CLEANUP_DEFERRED_ON_SUCCESS;
+				InsertCleanupRecordInCurrentTransaction(CLEANUP_SHARD_PLACEMENT,
```

Comment:
It's a little unexpected to me that this appears in DropShardList, which now doesn't actually drop the shard list. What do you think of introducing a separate function for this with an appropriate name?

(e.g. `InsertDeferredDropCleanupRecordsForShards(List *shardList)`)

Reply:
Makes sense. I have refactored the code. One slight change now is that I mark for deferred drop / drop shard first and then delete metadata as we need to traverse the MD twice.